Make account changes email bilingual

AntoineAugusti opened the pull request:
Fixes https://github.com/cds-snc/notification-api/issues/1016

Related previous PR https://github.com/cds-snc/notification-api/pull/459

Local demo:
![Screen Shot 2020-09-16 at 16 25 03](https://user-images.githubusercontent.com/295709/93388766-3557eb00-f839-11ea-8811-39b758479a11.png)

AntoineAugusti commented on app/user/rest.py:
```diff
@@ -164,8 +164,7 @@ def update_user_attribute(user_id):
                 'name': user_to_update.name,
                 'servicemanagername': updated_by.name,
-                'email address': user_to_update.email_address,
-                'change_type': update_dct_to_str(update_dct)
```
Discovered that the `change_type` is not used in the templates associated with the method.

Can you double-check? Template IDs should be:
```
TEAM_MEMBER_EDIT_EMAIL_TEMPLATE_ID = 'c73f1d71-4049-46d5-a647-d013bdeca3f0'
TEAM_MEMBER_EDIT_MOBILE_TEMPLATE_ID = '8a31520f-4751-4789-8ea1-fe54496725eb'
```

willeybryan replied:
Yes!
To find the templates you can go to Service 0, which handles GC Notify's Notifications
https://notification.alpha.canada.ca/services/d6aa2c68-a2d9-4437-ab19-3ae8eb202553/templates/[TEMPLATE_ID]

EDIT_EMAIL has variables ((servicemanagername)) ((name)) and ((email address)), So rather then saying your email was changed it says the new email they changed it too, this is funny because it will go to the new email.

EDIT_MOBILE has only ((servicemanagername)) and of course the number it goes it new phone number. 

So they are clear of the change type and this particular problem.

AntoineAugusti replied:
I exactly did that but wanted to make sure I did not miss anything 💯

AntoineAugusti commented on app/utils.py:
```diff
@@ -127,12 +127,36 @@ def email_address_is_nhs(email_address):
-        str += "- {}".format(key.replace("_", " "))
-        str += "\n"
-    return str
+        try:
+            key_name = values[lang][key]
+        except KeyError:
```
My implementation is: have values in both languages that I'm almost certain they will be used after carefully looking at the code, but provide a fallback if I missed keys.

The fallback is English attribute name + replace `_` to `_` (`email_address` to `email address`). The previous code was only doing this part.

AntoineAugusti commented on migrations/versions/0310_account_change_type.py:
```diff
@@ -0,0 +1,50 @@
+template = {
+    'id': current_app.config['ACCOUNT_CHANGE_TEMPLATE_ID'],
+    'subject': 'Account information changed | Informations de compte modifiées',
+    'content_lines': [
+        'Your GC Notify user account information was changed on ((base_url)).',
+        '',
+        'Updated information: ((change_type_en))',
+        '',
+        'If you did not make this change, [contact us](((contact_us_url)) "contact us") immediately.',
+        '',
+        '___',
+        '',
+        "Les informations de votre compte d''utilisateur ont été modifiées : ((base_url)).",
+        '',
+        "Informations mises à jour : ((change_type_fr))",
+        '',
+        "Si vous n''avez pas effectué ce changement, [communiquez avec nous](((contact_us_url)) \"communiquez avec nous\") immédiatement.",
+    ]
```
Used the template on staging, it was different in the local database.

Notable change is `change_type_en` + `change_type_fr`

brdunfield replied:
You used the staging template here? Can you confirm if this is the same as prod or not? If not, we'll want the prod one here.

AntoineAugusti replied:
Confirmed it's the same in staging and prod 👌

AntoineAugusti commented on tests/app/test_utils.py:
```diff
@@ -70,11 +70,19 @@ def test_midnight_n_days_ago(current_time, arg, expected_datetime):
         "email_address": "test@test.com",
-        "auth_type": "sms"
+        "auth_type": "sms",
+        "dummy_key": "nope",
```
Added this value to prove the fallback mechanism

AntoineAugusti commented on app/utils.py:
```diff
@@ -127,12 +127,36 @@ def email_address_is_nhs(email_address):
-def update_dct_to_str(update_dct):
-    str = "\n"
+def update_dct_to_str(update_dct, lang):
+    if lang not in ['EN', 'FR']:
+        raise NotImplementedError
```
Is there a cleaner way to do this on the API side? We don't have supported lang + translation files here.

brdunfield replied:
Correct. Right now translations typically happen by passing a key to the admin which handles the translating. The API itself doesn't do enough translation to have a system for it.

Unless we passed these values from the admin (which could be a good idea given the way our logic is laid out, but would be a more complicated fix). Otherwise, this would be the best way.

brdunfield commented on app/utils.py:
```diff
@@ -127,12 +127,36 @@ def email_address_is_nhs(email_address):
+        raise NotImplementedError
+
+    values = {
+        'EN': {
```
Have a look in rest.py - there's a set_permissions() function there that also sends a change alert when a users' service permissions change. There may be some extra values needed for that.

AntoineAugusti replied:
Here is the current email sent when changing permissions. I guess it's not ideal but I don't see how we could improve it easily. The 🇫🇷   part is correct from a copywriting perspective

![Screen Shot 2020-09-17 at 11 39 47](https://user-images.githubusercontent.com/295709/93494032-91268080-f8da-11ea-9fa3-25a794dc9b85.png)

brdunfield commented:
Paul just opened up this ticket on the admin repo:
https://github.com/cds-snc/notification-admin/issues/761

We could potentially roll this into this PR? Should be a pretty small fix.

AntoineAugusti commented:
Fixed https://github.com/cds-snc/notification-admin/issues/761 in https://github.com/cds-snc/notification-api/pull/1128/commits/1dc35c956963eba59db93e415356d1ba8ec36045

brdunfield commented:
I am not sure if it was like this before this PR or not, but when I make a chance as a service manager to one of my team member's accounts (I changed their email) - the team member gets both the email that someone changed their email, and an email stating THEY changed their email.

AntoineAugusti commented:
> I am not sure if it was like this before this PR or not, but when I make a change as a service manager to one of my team member's accounts (I changed their email) - the team member gets both the email that someone changed their email, and an email stating THEY changed their email.

It's not a regression but I fixed it in https://github.com/cds-snc/notification-api/pull/1128/commits/6cd54178a0d78d8589e7224f1b5a42eb6c3aaa03. Reproduced it locally and fixed it

brdunfield reviewed (approved):
LGTM, it's working for me locally as intended 👍 🦑

AntoineAugusti commented on tests/app/user/test_rest.py:
```diff
@@ -372,6 +372,7 @@ def test_post_user_attribute_with_updated_by(
     assert json_resp['data'][user_attribute] == user_value
 
     if arguments:
+        assert mock_persist_notification.call_count == 1
```
Was 2 calls before making the change. One "something on your profile changed" and another "an admin changed a thing". Making sure only the "admin changed a thing" is sent

AntoineAugusti commented:
Thanks for your careful and speedy review @brdunfield! I'll merge as soon as the build is green :shipit:
